File: binding/ruby/lib/milter/client/command-line.rb
# Copyright (C) 2010-2021  Sutou Kouhei <kou@clear-code.com>
#
# This library is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this library.  If not, see <http://www.gnu.org/licenses/>.

require "optparse"
require "ostruct"
require "syslog"

module Milter
  class Client
    class CommandLine
      attr_reader :options, :configuration, :option_parser
      def initialize(options={})
        setup_options
        setup_configuration(options)
        @option_parser = OptionParser.new(banner)
        yield(self) if block_given?
        setup_option_parser
      end

      def parse(argv=nil)
        begin
          @option_parser.parse!(argv || ARGV)
        rescue OptionParser::ParseError
          puts($!.message)
          exit(false)
        rescue
          puts("#{$!.class}: #{$!.message}")
          puts($@)
          exit(false)
        end
      end

      def run(argv=nil)
        parse(argv)
        client = Milter::Client.new
        client.on_error do |_client, error|
          Milter::Logger.error("[client][error] #{error.message}")
        end
        @configuration.setup(client)
        client.event_loop = client.create_event_loop(true)
        yield(client, options) if block_given?
        client.listen
        client.drop_privilege
        client.daemonize if @configuration.milter.daemon?
        setup_signal_handler(client) if @configuration.milter.handle_signal?
        client.run
      end

      def name
        @configuration.milter.name
      end

      private
      def setup_options
        @options = OpenStruct.new
      end

      def setup_configuration(options)
        @configuration = Configuration.new
        @configuration.milter.name = options[:name] || File.basename($0, ".*")
      end

      def setup_option_parser
        setup_common_options
      end

      def banner
        "Usage: %s [options]" % @configuration.milter.name
      end

      def setup_common_options
        setup_basic_options
        setup_milter_options
        setup_logger_options
      end

      def setup_basic_options
        @option_parser.separator ""
        @option_parser.separator "Basic options"

        @option_parser.on("--help", "Show this message.") do
          puts @option_parser
          exit(true)
        end

        @option_parser.on("--library-version",
                          "Show milter library version.") do
          puts Milter::VERSION.join(".")
          exit(true)
        end

        @option_parser.on("-cFILE", "--configuration=FILE",
                          "Load configuration from FILE") do |path|
          loader = ConfigurationLoader.new(@configuration)
          loader.load(path)
        end

        @option_parser.on("-eENV", "--environment=ENV",
                          "Set milter environment as ENV",
                          "(#{@configuration.environment})") do |env|
          @configuration.environment = env
        end
      end

      def setup_milter_options
        @option_parser.separator ""
        @option_parser.separator "milter options"

        milter_conf = @configuration.milter
        security_conf = @configuration.security
        @option_parser.on("-s", "--connection-spec=SPEC",
                          "Specify connection spec as [SPEC].",
                          "(#{milter_conf.connection_spec})") do |spec|
          milter_conf.connection_spec = spec
        end

        @option_parser.on("--[no-]daemon",
                          "Run as a daemon process.",
                          "(#{milter_conf.daemon?})") do |run_as_daemon|
          milter_conf.daemon = run_as_daemon
        end

        @option_parser.on("--pid-file=FILE",
                          "Write process ID to FILE.") do |pid_file|
          milter_conf.pid_file = pid_file
        end

        loader = ConfigurationLoader::MilterConfigurationLoader.new(loader)
        available_statues = loader.available_fallback_statuses
        available_values = "available values: [#{available_statues.join(', ')}]"
        @option_parser.on("--fallback-status=STATUS",
                          available_statues,
                          "Specify status on error.",
                          available_values,
                          "(#{milter_conf.fallback_status})") do |status|
          milter_conf.fallback_status = status
        end

        @option_parser.on("--status-on-error=STATUS",
                          "Deprecated option.",
                          "Use --fallback-status instead.") do |status|
          milter_conf.fallback_status = status
        end

        @option_parser.on("--user=USER",
                          "Run as USER's process (need root privilege)",
                          "(#{security_conf.effective_user})") do |user|
          security_conf.effective_user = user
        end

        @option_parser.on("--group=GROUP",
                          "Run as GROUP's process (need root privilege)",
                          "(#{security_conf.effective_group})") do |group|
          security_conf.effective_group = group
        end

        @option_parser.on("--unix-socket-group=GROUP",
                          "Change UNIX domain socket group to GROUP",
                          "(#{milter_conf.unix_socket_group})") do |group|
          milter_conf.unix_socket_group = group
        end

        @option_parser.on("--unix-socket-mode=MODE",
                          "Change UNIX domain socket mode to MODE",
                          "(%o)" % milter_conf.unix_socket_mode) do |mode|
          milter_conf.unix_socket_mode = mode
        end

        @option_parser.on("--max-file-descriptors=NUM",
                          "Change maximum number of file descriptors to NUM",
                          Integer) do |num|
          milter_conf.max_file_descriptors = num
        end

        backends = Milter::ClientEventLoopBackend.values.collect do |value|
          value.nick
        end
        @option_parser.on("--event-loop-backend=BACKEND", backends,
                          "Use BACKEND as event loop backend.",
                          "available values: [#{backends.join(', ')}]",
                          "(#{milter_conf.event_loop_backend})") do |backend|
          milter_conf.event_loop_backend = backend
        end

        @option_parser.on("--n-workers=N",
                          Integer,
                          "Run with N workrs.",
                          "(#{milter_conf.n_workers})") do |n|
          raise OptionParser::InvalidArgument if n <= 0
          milter_conf.n_workers = n
        end

        @option_parser.on("--packet-buffer-size=SIZE",
                          Integer,
                          "Use SIZE as packet buffer size.",
                          "(#{milter_conf.packet_buffer_size})") do |size|
          raise OptionParser::InvalidArgument if size < 0
          milter_conf.packet_buffer_size = size
        end

        @option_parser.on("--max-pending-finished-sessions=N",
                          Integer,
                          "Don't hold over N pending finished sessions.",
                          "(#{milter_conf.max_pending_finished_sessions})") do |n|
          raise OptionParser::InvalidArgument if n <= 0
          milter_conf.max_pending_finished_sessions = n
        end

        @option_parser.on("--[no-]handle-signal",
                          "Handle SIGHUP, SIGINT and SIGTERM signals",
                          "(#{milter_conf.handle_signal?})") do |boolean|
          milter_conf.handle_signal = boolean
        end

        @option_parser.on("--maintenance-interval=N",
                          "Run maintenance callback after each N sessions",
                          "(#{milter_conf.maintenance_interval})") do |interval|
          milter_conf.maintenance_interval = interval
        end

        @option_parser.on("--[no-]run-gc-on-maintain",
                          "Run GC in each maintenance callback",
                          "(#{milter_conf.run_gc_on_maintain?})") do |boolean|
          milter_conf.run_gc_on_maintain = boolean
        end
      end

      def setup_logger_options
        @option_parser.separator ""
        @option_parser.separator "Logging options"

        log_conf = @configuration.log
        level_names = Milter::LogLevelFlags.values.collect {|value| value.nick}
        level_names << "all"
        @option_parser.on("--log-level=LEVEL",
                          "Specify log level as LEVEL.",
                          "Select from [%s]." % level_names.join(", "),
                          "(#{ENV['MILTER_LOG_LEVEL'] || 'default'})") do |level|
          log_conf.level = level
        end

        @option_parser.on("--log-path=PATH",
                          "Specify log output path to PATH.",
                          "If PATH is '-', the standard output is used",
                          "(The standard output)") do |path|
          log_conf.path = path
        end

        @option_parser.on("--[no-]syslog",
                          "Use syslog",
                          "(#{log_conf.use_syslog?})") do |bool|
          log_conf.use_syslog = bool
        end

        facilities = Syslog.constants.find_all do |name|
          /\ALOG_/ =~ name.to_s
        end.collect do |name|
          name.to_s.gsub(/\ALOG_/, "").downcase
        end
        available_values = "available values: [#{facilities.join(', ')}]"
        @option_parser.on("--syslog-facility=FACILITY", facilities,
                          "Use FACILITY as syslog facility.",
                          "(#{log_conf.syslog_facility})",
                          available_values) do |facility|
          log_conf.syslog_facility = facility
        end

        @option_parser.on("--verbose",
                          "Show messages verbosely.",
                          "Alias of --log-level=all.") do
          log_conf.level = "all"
        end
      end

      def setup_signal_handler(client)
        # FIXME: This is just a workaround to handle signals within 0.5 seconds.
        # We should use more clever approach instead of this.
        if client.event_loop_backend == Milter::ClientEventLoopBackend::LIBEV
          client.event_loop.add_timeout(0.5) do
            true
          end
        end

        if client.n_workers == 0
          setup_signal_handler_reload(client)
          setup_signal_handler_reopen_log(client)
          setup_signal_handler_shutdown(client)
        end

        client.signal_connect("worker-created") do
          setup_signal_handler_reload(client)
          setup_signal_handler_reopen_log(client)
        end

        client.signal_connect("workers-created") do
          setup_signal_handler_reload(client) do
            Process.kill(:HUP, *client.worker_pids)
          end
          setup_signal_handler_reopen_log(client) do
            Process.kill(:USR1, *client.worker_pids)
          end
          setup_signal_handler_shutdown(client)
        end
      end

      def setup_signal_handler_reload(client)
        add_signal_handler(client, :HUP) do
          client.reload
          yield(client) if block_given?
        end
      end

      def setup_signal_handler_reopen_log(client)
        add_signal_handler(client, :USR1) do
          Milter::Logger.default.reopen
          yield(client) if block_given?
        end
      end

      def setup_signal_handler_shutdown(client)
        add_signal_handler(client, :INT,  :once => true) {client.shutdown}
        add_signal_handler(client, :TERM, :once => true) {client.shutdown}
      end

      def add_signal_handler(client, signal, options={})
        notification_message = "N"
        notification_in, notification_out = IO.pipe
        notification_in_channel = GLib::IOChannel.new(notification_in)

        event_loop = client.event_loop
        event_loop.watch_io(notification_in_channel, GLib::IOCondition::IN) do
          notification_in.read(notification_message.size)
          yield
          if options[:once]
            notification_in.close
            notification_out.close
            continue_p = false
          else
            continue_p = true
          end
          continue_p
        end

        default_handler = trap(signal) do
          notification_out.write(notification_message)
          notification_out.flush
          trap(signal, default_handler) if options[:once]
        end
      end
    end
  end
end

File: binding/python/gi/overrides/MilterClient/command_line.py
# Copyright (C) 2022  Sutou Kouhei <kou@clear-code.com>
#
# This library is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this library.  If not, see <http://www.gnu.org/licenses/>.

import argparse
import configparser
import contextlib
import os
import resource
import sys
import syslog

import gi._ossighelper
import gi.module

import milter.core
from .client import Client

MilterClient = gi.module.get_introspection_module("MilterClient")

class CommandLine(object):
    AVAILABLE_EVENT_LOOP_BACKENDS = [
        enum.value_nick
        for enum
        in MilterClient.ClientEventLoopBackend.__enum_values__.values()
    ]

    AVAILABLE_FALLBACK_STATUSES = [
        "accept",
        "reject",
        "temporary-failure",
        "discard",
    ]

    def __init__(self, name=None, version=None, **kwargs):
        self.name = name or os.path.splitext(os.path.basename(sys.argv[0]))[0]
        self.version = version
        self.parser = argparse.ArgumentParser(**kwargs)
        self._setup_arguments()

    @contextlib.contextmanager
    def run(self, argv=None):
        args = self._parse(argv)
        self._apply_args(args)
        client = Client()
        def on_error(_client, error):
            milter.core.Logger.error(f"[client][error] {type(error)}: {error}")
        self._setup_client(client, args)
        client.connect("error", on_error)
        client.set_event_loop(client.create_event_loop(True))
        yield client, args
        client.listen()
        client.drop_privilege()
        if args.daemon:
            client.daemonize()
        with gi._ossighelper.register_sigint_fallback(lambda: client.shutdown()):
            with gi._ossighelper.wakeup_on_signal():
                client.run()

    def _setup_arguments(self):
        self._setup_basic_arguments()
        self._setup_milter_arguments()
        self._setup_logger_arguments()

    class ShowLibraryVersionAction(argparse.Action):
        def __call__(self, *args, **kwargs):
            print(milter.core.VERSION)
            sys.exit(0)

    class ConfigurationLoadAction(argparse.Action):
        def __call__(self, parser, namespace, path, option_string):
            namespace.config_path = path
            if not hasattr(namespace, "config"):
                namespace.config = configparser.ConfigParser()
            config = namespace.config
            with open(path) as input:
                config.read_file(input, path)

            def set_config(key, section, option,
                           getter=config.get,
                           available_values=None):
                if not section in config:
                    return
                current_value = getattr(namespace, key)
                setattr(namespace,
                        key,
                        getter(section, key, fallback=current_value))

            set_config("environment", "basic", "environment")
            set_config("connection_spec", "milter", "connection_spec")
            set_config("daemon", "milter", "daemon", config.getboolean)
            set_config("pid_file", "milter", "pid_file")
            set_config("fallback_status",
                       "milter",
                       "fallback_status",
                       available_values=CommandLine.AVAILABLE_FALLBACK_STATUSES)
            set_config("effective_user", "milter", "user")
            set_config("effective_group", "milter", "group")
            set_config("unix_socket_group", "milter", "unix_socket_group")
            set_config("unix_socket_mode", "milter", "unix_socket_mode")
            set_config("max_file_descriptors", "milter", "max_file_descriptors")
            set_config("event_loop_backend",
                       "milter",
                       "event_loop_backend",
                       available_values=CommandLine.AVAILABLE_EVENT_LOOP_BACKENDS)
            set_config("log_level", "log", "level")
            set_config("log_path", "log", "path")
            set_config("use_syslog", "log", "use_syslog", config.getboolean)
            set_config("syslog_facility", "log", "syslog_facility")

    def _setup_basic_arguments(self):
        basic = self.parser.add_argument_group("Basic", "Basic options")
        if self.version:
            basic.add_argument("--version",
                               action="version",
                               help="Show version",
                               version=f"%(prog)s {self.version}")
        basic.add_argument("--library-version",
                           action=self.ShowLibraryVersionAction,
                           help="Show milter library version",
                           nargs=0)
        basic.add_argument("-c", "--configuration",
                           action=self.ConfigurationLoadAction,
                           help="Load configuration from FILE",
                           metavar="FILE")
        basic.add_argument("-e", "--environment",
                           default=os.environ.get("MILTER_ENV", "development"),
                           dest="environment",
                           help="Set milter environment as ENVIRONMENT.\n" +
                           "(default: %(default)s)",
                           metavar="ENVIRONMENT")

    def _setup_milter_arguments(self):
        milter = self.parser.add_argument_group("milter", "milter options")
        milter.add_argument("--name",
                            default=self.name,
                            dest="name",
                            help="Specify name as NAME.\n" +
                            "(default: %(default)s)",
                            metavar="NAME")
        milter.add_argument("-s", "--connection-spec",
                            default="inet:20025",
                            dest="connection_spec",
                            help="Specify connection spec as SPEC.\n" +
                            "(default: %(default)s)",
                            metavar="SPEC")
        if hasattr(argparse, "BooleanOptionalAction"):
            milter.add_argument("--daemon",
                                action=argparse.BooleanOptionalAction,
                                dest="daemon",
                                help="Run as a daemon process")
        milter.add_argument("--pid-file",
                            dest="pid_file",
                            help="Write process ID to FILE",
                            metavar="FILE")
        milter.add_argument("--fallback-status",
                            choices=self.AVAILABLE_FALLBACK_STATUSES,
                            default="accept",
                            dest="fallback_status",
                            help="Use STATUS as fallback status.\n" +
                            "(available values: %(choices)s)\n" +
                            "(default: %(default)s)",
                            metavar="STATUS")
        milter.add_argument("--user",
                            dest="effective_user",
                            help="Run as USER's process (need root privilege)",
                            metavar="USER")
        milter.add_argument("--group",
                            dest="effective_group",
                            help="Run as GROUP's process (need root privilege)",
                            metavar="GROUP")
        milter.add_argument("--unix-socket-group",
                            dest="unix_socket_group",
                            help="Change UNIX domain socket group to GROUP",
                            metavar="GROUP")
        milter.add_argument("--unix-socket-mode",
                            dest="unix_socket_mode",
                            help="Change UNIX domain socket mode to MODE",
                            metavar="MODE")
        milter.add_argument("--max-file-descriptors",
                            dest="max_file_descriptors",
                            help="Change maximum number of file descriptors to N",
                            metavar="N",
                            type=int)
        milter.add_argument("--event-loop-backend",
                            choices=self.AVAILABLE_EVENT_LOOP_BACKENDS,
                            default="default",
                            dest="event_loop_backend",
                            help="Use BACKEND as event loop backend.\n" +
                            "(available values: %(choices)s)\n" +
                            "(default: %(default)s)",
                            metavar="BACKEND")

        # TODO: See Milter::Client::CommandLine#setup_milter_options

    def _setup_logger_arguments(self):
        logger = self.parser.add_argument_group("Logger", "Logger options")
        level_names = [
            value.first_value_nick
            for value
            in milter.core.LogLevelFlags.__flags_values__.values()
        ]
        level_names += ["all"]
        logger.add_argument("--log-level",
                            default="default",
                            dest="log_level",
                            help="Specify log level as LEVEL.\n" +
                            "You can set levels by LEVEL1|LEVEL2|...." +
                            "You can add one or more levels " +
                            "to the default levels by +LEVEL1|+LEVEL2." +
                            "You can remove one or more levels " +
                            "from the default levels by -LEVEL1|-LEVEL2." +
                            f"(available levels: {', '.join(level_names)})\n" +
                            "(default: %(default)s)",
                            metavar="LEVEL")
        logger.add_argument("--log-path",
                            default="-",
                            dest="log_path",
                            help="Specify log output path as PATH.\n" +
                            "If PATH is '-', the standard output is used.\n" +
                            "(default: %(default)s)",
                            metavar="PATH")
        if hasattr(argparse, "BooleanOptionalAction"):
            logger.add_argument("--syslog",
                                default=False,
                                action=argparse.BooleanOptionalAction,
                                dest="use_syslog",
                                help="Use syslog")
        no_facility_names = [
            "LOG_EMERG",
            "LOG_ALERT",
            "LOG_CRIT",
            "LOG_ERR",
            "LOG_WARNING",
            "LOG_NOTICE",
            "LOG_INFO",
            "LOG_DEBUG",
            "LOG_PID",
            "LOG_CONS",
            "LOG_NDELAY",
            "LOG_ODELAY",
            "LOG_NOWAIT",
            "LOG_PERROR",
        ]
        syslog_facilities = [
            key.replace("LOG_", "", 1).lower()
            for key
            in syslog.__dict__
            if key.startswith("LOG_") and not key in no_facility_names
        ]
        logger.add_argument("--syslog-facility",
                            choices=syslog_facilities,
                            default="mail",
                            dest="syslog_facility",
                            help="Use FACILITY as syslog facility.\n" +
                            "(available values: %(choices)s).\n" +
                            "(default: %(default)s)",
                            metavar="FACILITY")
        logger.add_argument("--verbose",
                            action="store_const",
                            const="all",
                            dest="log_level",
                            help="Show messages verbosely.\n" +
                            "Alias of --log-level=all.")

    def _parse(self, argv=None):
        args = self.parser.parse_args(argv)
        if not hasattr(args, "config"):
            args.config = configparser.ConfigParser()
        return args

    def _apply_args(self, args):
        milter.core.Logger.get_default().target_level = args.log_level
        milter.core.Logger.get_default().path = args.log_path

    def _setup_client(self, client, args):
        client.set_connection_spec(args.connection_spec)
        if args.pid_file:
            client.set_pid_file(args.pid_file)
        if args.fallback_status:
            client.fallback_status = args.fallback_status
        if args.use_syslog:
            client.start_syslog(args.name, args.syslog_facility)
        if args.effective_user:
            client.set_effective_user(args.effective_user)
        if args.effective_group:
            client.set_effective_user(args.effective_group)
        if args.unix_socket_group:
            client.set_unix_socket_group(args.unix_socket_group)
        mode = args.unix_socket_mode
        if mode:
            if mode.startswith("0"):
                mode = int(mode, 8)
            else:
                mode = int(mode)
            client.set_unix_socket_mode(mode)
        max_file_descriptors = args.max_file_descriptors
        if max_file_descriptors:
            max_file_descriptors = int(max_file_descriptors)
            if max_file_descriptors > 0:
                resource.setrlimit(resource.RLIMIT_NOFILE,
                                   (max_file_descriptors, max_file_descriptors))
        if args.event_loop_backend:
            client.set_event_loop_backend(
                getattr(MilterClient.ClientEventLoopBackend,
                        args.event_loop_backend.upper()))
